nexusclient/client.go

```diff
@@ -0,0 +1,725 @@
+// It has two implementations: [OperationCompletionSuccessful] and [OperationCompletionUnsuccessful].
+type OperationCompletion interface {
+	io.Closer
+	applyToHTTPRequest(*http.Request, *Client) error
+}
+
+// Input for [nexusclient.Client.CompletionOperation] to deliver successful operation results.
+type OperationCompletionSuccessful struct {
+	// Header to send in the completion request.
+	Header http.Header
+	// Body to send in the completion HTTP request.
+	// If it implements `io.Closer` it will automatically be closed by the client
+	Body io.Reader
+}
+
+// Input for [nexusclient.Client.CompletionOperation] to deliver unsuccessful operation results.
+type OperationCompletionUnsuccessful struct {
+	// Header to send in the completion request.
+	Header http.Header
+	// State of the operation, should be failed or canceled.
+	State nexusapi.OperationState
+	// Failure object to send with the completion.
+	Failure *nexusapi.Failure
+}
+
+// NewJSONSuccessfulOperationCompletion constructs an [OperationCompletionSuccessful] from a JSONable value.
+// Marhsals the provided value to JSON using [nexusapi.DefaultMarshaler].
+func NewJSONSuccessfulOperationCompletion(v any) (*OperationCompletionSuccessful, error) {
+	b, err := nexusapi.DefaultMarshaler(v)
+	if err != nil {
+		return nil, err
+	}
+
+	header := make(http.Header, 1)
+	header.Set(nexusapi.HeaderContentType, nexusapi.ContentTypeJSON)
+
+	return &OperationCompletionSuccessful{
+		Header: header,
+		Body:   io.NopCloser(bytes.NewReader(b)),
+	}, nil
+}
+
+func (c *OperationCompletionSuccessful) applyToHTTPRequest(request *http.Request, client *Client) error {
+	if c.Header != nil {
+		request.Header = c.Header.Clone()
+	}
+	request.Header.Set(nexusapi.HeaderOperationState, string(nexusapi.OperationStateSucceeded))
+	if closer, ok := c.Body.(io.ReadCloser); ok {
+		request.Body = closer
+	} else {
+		request.Body = io.NopCloser(c.Body)
+	}
+	return nil
+}
+
+// Close implements the io.Closer interface.
+func (c *OperationCompletionSuccessful) Close() error {
+	if closer, ok := c.Body.(io.Closer); ok {
+		return closer.Close()
+	}
+	return nil
+}
+
+func (c *OperationCompletionUnsuccessful) applyToHTTPRequest(request *http.Request, client *Client) error {
+	if c.Header != nil {
+		request.Header = c.Header.Clone()
+	}
+	request.Header.Set(nexusapi.HeaderOperationState, string(c.State))
+	request.Header.Set(nexusapi.HeaderContentType, nexusapi.ContentTypeJSON)
+
+	b, err := client.Options.Marshaler(c.Failure)
+	if err != nil {
+		return err
+	}
+
+	request.Body = io.NopCloser(bytes.NewReader(b))
+	return nil
+}
+
+// Close implements the io.Closer interface.
+func (c *OperationCompletionUnsuccessful) Close() error {
+	return nil
+}
+
+// DeliverCompletion delivers the result of a completed asynchronous operation to the provided URL.
+func (c *Client) DeliverCompletion(ctx context.Context, url string, completion OperationCompletion) error {
```

Comment:
I think this is a server side thing not a client side thing. It shouldn't be on the Nexus client that operation invokers use. Arguably this shows that maybe we don't need separate package for `nexusclient` and `nexusserver` but just `nexus`.